src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -321,6 +336,13 @@ static ucs_status_t uct_tcp_sockcm_ep_progress_send(uct_tcp_sockcm_ep_t *cep)
                   cep->fd, events, ucs_status_string(status));
     }
 
+    if (cep->state & UCT_TCP_SOCKCM_EP_SERVER_REJECT_SENT) {
+        /* return an error status so that if the reject message was sent in
+         * batches, the event handler would close the server's internal endpoint
+         * after completing the send */
+        status = UCS_ERR_REJECTED;
```

Comment:
why not return OK?

Reply:
so that if the reject message is sent in batches, not all at once, the event handler would destroy the server's ep in the event handler in uct_tcp_sockcm_ep_handle_event_status() after the reject is sent. 
we can't rely only on checking the reject_sent flag in the event_handler since if the reject was sent in one call from the conn_req_cb, the listener's reject function would already destroy the ep which means we cant check the flag in the event_handler.

Reply:
but uct_tcp_sockcm_ep_handle_event_status() is called only in case send fails, right? so why not destroy the ep in case of end failure?

Reply:
we also need to destroy the ep if send passes (after reject completes). it's an internal server ep. uct_ep_desrtoy won't be called on it.

Reply:
can't we destroy the ep by checking status? 
we should not return ERR_REJECTED from uct_tcp_sockcm_ep_progress_send() because it doesn't make sense to return error when the function did not really fail

Reply:
ww can't check the ep's status in the event_handler since the listener's reject function might have already destroyed this ep. 
i agree that this isn't optimal. other way is destroy the ep in the uct_tcp_sockcm_ep_progress_send() function if reject finished but i guess that it's wrong too since this is a sending function, it shouldn't destroy/create anything.

Reply:
another idea - how about i remove the uct_tcp_sockcm_ep_progress_send() from the listener's reject function? that way, the reject message will be sent only from the event_handler and will always exist after the send completes. 
it will add a slight delay in the reject function but very small i guess.

Reply:
i think we should destroy the ep from uct_tcp_sockcm_ep_progress_send() function, after checking uct_tcp_sockcm_ep_is_tx_rx_done() and the status flag. before destroying the ep we are removing it from the event handler so should not get events. it's not really different from a destroy operation for any other reason.

Reply:
Removing the ep from the event handler is part of the ep's destroy function.
Didn't think it was ok to destroy an ep from a sending function... But ok, will do it there.